File: fnc/examples/chapter09/lsfitexp.m
%% 
% Let's approximate $e^x$ over the interval $[-1,1]$. We can sample it at,
% say, 40 points, and find the best-fitting straight line to that data. 
t = linspace(-1,1,40)';
y = exp(t);
plot(t,y,'.')
V = [t.^0 t];
c = V\y
p = @(t) c(1) + c(2)*t;
hold on, fplot(p,[-1 1])
title('Least-squares fit to data from exp(x)')    % ignore this line
xlabel('x'), ylabel('f(x) or p(x)')    % ignore this line

%%
% There's nothing special about 40 points. By choosing more we get closer
% to the ``true'' $e^x$.
t = linspace(-1,1,200)';
y = exp(t);
V = [t.^0 t];
c = V\y

%%
t = linspace(-1,1,1000)';
y = exp(t);
V = [t.^0 t];
c = V\y

%%
% It's quite plausible that the coefficients of the best-fit line are
% approaching a limit as the number of nodes goes to infinity. 

File: julia/scripts/chapter09/lsfitexp.jl

t = range(-1,stop=1,length=40)
y = exp.(t)

using Plots
plot(t,y,m=(3,:o),label="function")

V = [ ti^j for ti in t, j=0:1 ]
c = V\y
plot!(t->c[1]+c[2]*t,-1,1,label="fit",
    xaxis=("x"),yaxis=("value"),title="Least squares fit of exp(x)",leg=:bottomright)

t = range(-1,stop=1,length=200)
y = exp.(t)

plot(t,y,m=(2,:o),label="function")

V = [ ti^j for ti in t, j=0:1 ]
c = V\y
plot!(t->c[1]+c[2]*t,-1,1,label="fit",
    xaxis=("x"),yaxis=("value"),title="Least squares fit of exp(x)",leg=:bottomright)

t = range(-1,stop=1,length=1000)
y = exp.(t)

plot(t,y,m=(1,:o),label="function")

V = [ ti^j for ti in t, j=0:1 ]
c = V\y
plot!(t->c[1]+c[2]*t,-1,1,label="fit",
    xaxis=("x"),yaxis=("value"),title="Least squares fit of exp(x)",leg=:bottomright)
